design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+
+```python
+def requirements(self):
+    self.requires(ref, headers=True, libs=True, run=None, visible=True, …)
+```
+
+There are defaults defined, so the same previous Conan 1.X syntax of class attribute ``requires ="ref"`` and ``self.requires(ref)`` is still possible.
+
+Reminder, following the GNU standard naming:
+
+- Build context: The machine in which the current Conan process and build is running. If we are cross-compiling in a Windows box for an app that will run in a Linux RaspberryPI environment, the build context is the Windows machine.
+- Host context: The machine that will run the apps and libraries being built, in this example the host context is the Linux running in the RaspberryPI
+
+
+The traits are:
+
+- ``headers`` (default True): This trait indicates that there are headers that are going to be #included from this package at compile time. The dependency will be in the host context.
+- ``libs`` (default True): The dependency contains some library or artifact that will be used at link time of the consumer. The dependency will be in the host context. This trait will be true for direct shared and static libraries, but could be false for indirect static libraries that are consumed via a shared library.
+- ``build`` (default False): This dependency is a build tool, an application or executable, like cmake, that is used exclusively at build time, it is not linked/embedded into binaries, and will be in the build context.
+- ``run`` (default None): This dependency contains some executables, either apps or shared libraries that need to be available to execute (typically in the path, or other system env-vars). By default is None, indicating that Conan can try to deduce if the dependency is needed to execute (if ``options.shared`` is True). This trait can be True for ``build=False``, in that case, the package will contain some executables that can run in the host system when installing it, typically like an end-user application. This trait can be True for ``build=True``, the package will contain executables that will run in the build context, typically while being used to build other packages.
+- ``visible`` (default True): This ``require`` will be propagated downstream, even if it doesn’t propagate ``headers``, ``libs`` or ``run`` traits. Requirements that propagate downstream can cause version conflicts. This is by default True, because in most cases, having 2 different versions of the same library in the same dependency graph is at least complicated, if not directly violating ODR or causing linking errors.
+- ``transitive_headers`` (default None): The headers of the dependency will be visible downstream or not. The default None allows Conan to auto detect this trait, for example, if the current package is a ``header-only`` one, and it depends on another library (header only, or static/shared), the headers of the transitive dependency must be available and used in the ``-I<includedirs>`` compilation downstream.
```

Comment:
What are allowed values here? Is this a boolean value (so, `None`, `True` and `False`) or is it expected to place here the array of folder/file paths that should be transitive?

Reply:
It is expected to be boolean, None, True, False. 
Making it that more fine grained will probably be too complex and have a marginal value, maybe even an anti-pattern (it is very difficult to model the dependency internal headers requirements, you think that you would only be propagating ``include/poco/thread`` headers to your consumers, but those headers in turn depend on ``include/poco/system`` that in turn depends on other headers.

Reply:
I agree and I thought that the expected type is the boolean, but since it's not explicitly stated, I had to verify that. Maybe add some info that all traits are expected to be boolean? Or will there be a trait that is not a boolean?

Reply:
Probably the ``package_id_mode`` will not be boolean, but for all others, yes, they should be (that allows much simpler and systematic implementations later). I will add a note in the doc clarifying this.